### DISCUSSION TITLE
How to write the stress field into vtk file？

### DISCUSSION THREAD

**Discussion by zyhsjtu** [Ferrite-FEM/Ferrite.jl#912] [General]
How to write the stress field into vtk file.
I add the following codes and calculate the strain tensor and stress at quadrature point，
  `

    struct StateVariable{T}
       ε::SymmetricTensor{2,3,T} # strain in each quadrature point
      σ::SymmetricTensor{2,3,T} # stress in each quadrature point
      end
     function StateVariable()
          return StateVariable(zero(SymmetricTensor{2, 3}),zero(SymmetricTensor{2, 3}))
    end
`

Then I write the follow codes to extract the stress field and write it into vtk but failed.
`

    function compute_stress(states, cellvalues::CellVectorValues{dim, T}, dh::DofHandler) where {dim, T}
          σ = [SymmetricTensor{2,3, Float64}[] for _ in 1:getncells(dh.grid)]
	  for (i, cell) in enumerate(CellIterator(dh))
		  state = @view states[:, i]
		  σ_cell = σ[i]
		  for q_point in 1:getnquadpoints(cellvalues)
			  push!(σ_cell, state[q_point].σ)
		  end
	  end
	  return σ
    end
     vtk_grid(filename, grid) do vtk
           vtk_point_data(vtk,  σ, "σ(MPa)")
        end

`

and the error information is:

`
ERROR: LoadError: DimensionMismatch: incorrect dimensions of input array.
`

So How should I do to write the stress field and how to project them into nodes?

**Comment by termi-official** [comment]
Quadrature data is usually projected to nodes with an L2 projection. Here you solve the problem $\int u*v = \int f*v$ where the expression on the right hand side is your quadrature data. For this we provide a small helper function `project` and the associated `L2Projector` struct, for which you can find an example in the [second tutorial, which is about postprocessing](https://ferrite-fem.github.io/Ferrite.jl/stable/examples/postprocessing).

**Reply by zyhsjtu** [reply]
Thanks for your quick response!! I have done it as follow, but still there some errors, could you give me some advices?
`

     projector = L2Projector(ip, grid);
     σ = compute_stress(states, cellvalues, dh);
     σ_projected = project(projector, σ, qr; project_to_nodes = false);
     vtk_grid(filename, grid) do vtk
       vtk_point_data(vtk, projector, σ_projected, "σ(MPa)")
    end
`

![image](https://github.com/Ferrite-FEM/Ferrite.jl/assets/82813702/8e204ca0-e46d-4e96-b113-73b007ecdc8f)

**Comment by KristofferC** [comment]
https://ferrite-fem.github.io/Ferrite.jl/stable/examples/plasticity/ should probably be updated to show how to do this properly instead of the slightly inelegant way that is currently used there.

**Reply by zyhsjtu** [reply]
Thanks for your suggestion!

**Reply by fredrikekre** [reply]
In https://ferrite-fem.github.io/Ferrite.jl/dev/tutorials/incompressible_elasticity/ there is volume averaging within the element.

**Reply by termi-official** [reply]
> https://ferrite-fem.github.io/Ferrite.jl/stable/examples/plasticity/ should probably be updated to show how to do this properly instead of the slightly inelegant way that is currently used there.

Yea, I will try to update the example with FerriteViz after release, so you can inspect the true stress fields.

**Reply by zyhsjtu** [reply]
> In https://ferrite-fem.github.io/Ferrite.jl/dev/tutorials/incompressible_elasticity/ there is volume averaging within the element.
Thanks for your reply, I was using the unupdated example earlier and didn't see  the code that output the stress field.
